bpo-47256: re module, limit the maximum capturing group to 1,073,741,823, increasing the depth of backtracking.

animalize opened the pull request:
Closes #91412.

serhiy-storchaka commented on Misc/NEWS.d/next/Library/2022-04-08-14-55-55.bpo-47256.X2mFpt.rst:
```diff
@@ -0,0 +1,2 @@
+:mod:`re` module, limit the maximum capturing group to 1,073,741,823,
```
It is only on 64-bit platforms. On 32-bit platforms it is still 4 times less.

serhiy-storchaka commented on Modules/_sre/sre.h:
```diff
@@ -73,9 +73,9 @@ typedef struct {
-    /* registers */
-    Py_ssize_t lastindex;
-    Py_ssize_t lastmark;
+    /* marks */
```
I think it is worth to move `match_all` and `must_advance` above this comment. They are not related to marks.

serhiy-storchaka commented on Modules/_sre/sre.h:
```diff
@@ -18,10 +18,10 @@
 #else
 # define SRE_MAXREPEAT ((SRE_CODE)PY_SSIZE_T_MAX)
-# define SRE_MAXGROUPS ((SRE_CODE)PY_SSIZE_T_MAX / SIZEOF_SIZE_T / 2)
+# define SRE_MAXGROUPS ((SRE_CODE)INT32_MAX / SIZEOF_INT / 2)
```
It should be `SIZEOF_VOID_P` in the denominator (it is the size of the `mark` item) and `PY_SSIZE_T_MAX` in the numerator, because the condition is no integer overflow in `state->mark = PyMem_New(const void *, pattern->groups * 2)`.

wjssz committed:
1. limit the maximum capturing group to 1,073,741,823

These types are kept as Py_ssize_t:
  - PatternObject.groups
  - MatchObject.lastindex
  - MatchObject.groups

- On 32 bit platform: 36 bytes, no change.  (msvc2022)
- On 64 bit platform: 72 bytes -> 64 bytes. (msvc2022/gcc9.4)

wjssz committed:
2. further reduce sizeof(SRE(match_context))

change the type of `SRE(match_context).jump` from Py_ssize_t to int

- On 32 bit platform: 36 bytes, no change.  (msvc2022)
- On 64 bit platform: 64 bytes -> 56 bytes. (msvc2022/gcc9.4)

And make the order of `DO_JUMPX` macro and `SRE(match_context)` struct consistent.

wjssz committed:
3. MARK_(PUSH|POP) macros compute marks size only once

DATA_STACK_(PUSH|POP) macros use the size multiple time, now computed value is propagated.

wjssz committed:
4. remove variable i

serhiy-storchaka commented on Modules/_sre/sre_lib.h:
```diff
@@ -610,20 +610,22 @@ SRE(match)(SRE_STATE* state, const SRE_CODE* pattern, int toplevel)
+                    /* state->lastmark is the highest valid index in the
+                    state->mark array.  If it is increased by more than 1,
+                    the intervening marks must be set to NULL to signal
+                    that these marks have not been encountered. */
```
Fix indentation.

wjssz committed:
Fix indentation
